MSC1943: Set default room as v3 (#1943)
matrix-org/matrix-doc: proposals/1943-v3-default-room-version.md

# Set v3 to be the default room version

Homeservers support multiple room versions, though one version is the default.

> All new rooms created by a server are of the default version.

Reviewer: the default is a recommendation; we can't change the behaviour of old servers.

Author: I don't see a contradiction here, whatever the default the server has, the description is accurate. I've made a change further down to clarify that the default is only a recommendation.

Reviewer 2: technically speaking `/createRoom` allows for specifying a room version which is not of the default, making the statement here untrue. However, I don't think this is the thing we need to get stuck on.

Reviewer: ok I think I misread this. All new rooms are inherently created by a server, so I just read it as "All new rooms are of the default version."

I think what you mean is "By default, new rooms created by a server are of the default version for that server", which is somewhat tautologous. I guess I'm just not sure this sentence adds clarity rather than confusion. Perhaps:

> Homeservers support multiple room versions. Each server has a default version which is used when creating new rooms.

> Room v3 contains support for the new state resolution algorithm and support

Reviewer: > Room v3 uses the new state resolution algorithm, and replaces event IDs with hashes, thereby closing a raft of security issues.

for event ids as hashes. Thereby closing a raft of security issues.

This MSC proposes changing the default from v1 to v3.

## Proposal

The default room should be updated to be room v3 from v1.

> Until we have greater confidence in the room upgrade UX v1 and v2 will continue

Reviewer 3: (non-blocking comment) Given that v2 is probably not widely used publicly, and would have been created by people who hopefully should know what they're doing, I'd be tempted to mark it as unstable right away.

Author: good idea

to be marked as 'stable'.

This means that new rooms will be created as v3 rooms, but existing rooms will

> not be prompted to upgrade. Separately we intend to deprecate v1 (and v2), but

Reviewer: ```suggestion
not prompt admins to upgrade. Separately we intend to deprecate v1 (and v2), but
```

> doing will come as part of a separate MSC

Reviewer: ```suggestion
doing so will come as part of a separate MSC.
```

## Tradeoffs

Room v4 will arrive very soon and ideally it will become the default room
version promptly. However in order for a room version to be the default there
must be sufficient take up in the federation so as to avoid locking out old
servers.

> That said v3 and v4 are very similar, the significant change being

Reviewer: V4 is not yet specced or even MSCed. I suggest removing this paragraph, but at least change it to record an expectation rather than a statement of fact.

[server key validity period](https://github.com/matrix-org/synapse/issues/4364)
and there is obvious value to the community to adopt v3 even in the
knowledge that v4 will surpass it shortly.

## Potential issues

If new rooms are to be v3 some old servers will not be able to participate

> since they will not support v3. For context v3 was introduced in Synapse 0.99.0

Reviewer: ```suggestion
since they will not support v3. For context, v3 was introduced in Synapse 0.99.0
```

on 5th Feb 2019.

Ignoring a long tail of of pre 0.30.0 servers (0.30.0 was released in
May 2018) ~2/3 servers support v3, and practically speaking ~90% of users are
on v3 compatible servers.

It does not seem unreasonable to expect Synapse admins to update to > 0.99.0

> server as a means to mitigate any difficulties caused by not support rooms v3.

Reviewer: ```suggestion
server as a means to mitigate any difficulties caused by not supporting v3 rooms.
```

> # Set v3 to be the default room version

Reviewer 4: If #1884 merges, then we should probably be setting default room to v4 instead.

Reviewer: unconvinced: it will take a while for v4 (whatever it ends up being) to roll out across the federation, whereas v3 has already been out for several weeks.

Author: I've changed my mind here fwiw. If we go to the trouble of implementing #1884 then I can't see sense in shipping v3 as default. Though agree with @richvdh comment that it pushes back the date that we can change default from v1.

# Set v3 to be the default room version

Homeservers support multiple room versions, though one version is the default.
All new rooms created by a server are of the default version.

Room v3 uses the new state resolution algorithm and support
for event ids as hashes. Thereby closing a raft of security issues.

Room v3 uses the new state resolution algorithm, and replaces event IDs with

> hashes, thereby closing a raft of security issues.

Reviewer 2: duplicated

This MSC proposes changing the recommended default from v1 to v3.

## Proposal

The default room should be updated to be room v3 from v1.

Until we have greater confidence in the room upgrade UX v1 will continue
to be marked as 'stable'.

By contrast, very few rooms will be of version v2, and this version can be
safely marked as 'unstable'.

This means that new rooms will be created as v3 rooms, but existing rooms will
not prompt admins to upgrade. Separately we intend to deprecate v1 (and v2), but

> doing so will come as part of a separate MSC.

Reviewer 2: this part confuses me: above seems to propose that v2 is being marked as `unstable` which is the route for deprecation

Author: Originally v2 was also to be deprecated at a later date, this is no longer the case, have changed accordingly.

Reviewer 2: It's definitely clearer now - spelling it out on line 19/20 might be worthwhile, but I think the concern I've raised here is resolved.

Author: Done
